### DISCUSSION TITLE
Tutorial How to use Go Whatsapp REST

### DISCUSSION THREAD

**Discussion by ibnux** [dimaskiddo/go-whatsapp-rest#40] [General]
Some of us maybe need Document for this, but i know the developer write this in his spare time 😄 
i just want to help, because its helping me make sending OTP with low cost

# INSTALATION
Read [README](https://github.com/dimaskiddo/go-whatsapp-rest/blob/master/README.md) File 😁 

# How To Use
first you need to see config
https://github.com/dimaskiddo/go-whatsapp-rest/blob/d6a00a048d98298c894d12e2853d9e1b0c9f71ab/config/production.yaml#L18

change that to any value.
now base64 that value with username, i use phone number

ex: `081234567:83e4060e-78e1-4fe5-9977-aeeccd46a2b8`
Result: `MDgxMjM0NTY3OjgzZTQwNjBlLTc4ZTEtNGZlNS05OTc3LWFlZWNjZDQ2YTJiOA==`

see route in here
https://github.com/dimaskiddo/go-whatsapp-rest/blob/d6a00a048d98298c894d12e2853d9e1b0c9f71ab/internal/route.go#L14-L28

https://github.com/dimaskiddo/go-whatsapp-rest/blob/d6a00a048d98298c894d12e2853d9e1b0c9f71ab/config/production.yaml#L11

# Authorization
This is only to get JWT token
you need to add **JWT_EXPIRATION_TIME_HOURS** to config for long expiration
```cli
curl --request GET \
  --url http://localhost:3000/api/v1/whatsapp/auth \
  --header 'Authorization: Basic MDgxMjM0NTY3OjgzZTQwNjBlLTc4ZTEtNGZlNS05OTc3LWFlZWNjZDQ2YTJiOA=='
```
## Result

```json
{
  "status": true,
  "code": 200,
  "message": "Success",
  "data": {
    "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJkYXRhIjoiMDgxMjM0NTY3OCIsImV4cCI6MTYzMzUyNzk3NH0.4FLvNzZw3nwt93TxmnynQkVPYCaoRzUhdWD8ri6Fy74"
  }
}
```
# Health Check
```
curl --request GET \
  --url http://127.0.0.1:3000/api/v1/whatsapp/health
```
## RESULT
```
{
  "status": true,
  "code": 200,
  "message": "Success"
}
```

# WhatsApp Login
```
curl --request POST \
  --url http://server.ibnux.net:3000/api/v1/whatsapp/login \
  --header 'Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJkYXRhIjoiMDgxMjM0NTY3OCIsImV4cCI6MTYzMzUyNzk3NH0.4FLvNzZw3nwt93TxmnynQkVPYCaoRzUhdWD8ri6Fy74' \
  --header 'Content-Type: application/x-www-form-urlencoded' \
  --data output=html
```

## RESULT
i run it using Insomnia Rest, and will show like this
scan it with your Whatsapp
<img width="332" alt="Screen Shot 2021-01-27 at 11 55 10" src="https://user-images.githubusercontent.com/811955/105944774-9371c600-6096-11eb-98a4-811643d1e381.png">

# Send Message
```
curl --request POST \
  --url http://127.0.0.1:3000/api/v1/whatsapp/send/text \
  --header 'Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJkYXRhIjoiMDgxMjM0NTY3OCIsImV4cCI6MTYzMzUyNzk3NH0.4FLvNzZw3nwt93TxmnynQkVPYCaoRzUhdWD8ri6Fy74' \
  --header 'Content-Type: application/x-www-form-urlencoded' \
  --data msisdn=6281122345678@s.whatsapp.net \
  --data 'message=halo kang'
```

## RESULT
```
{
  "status": true,
  "code": 200,
  "message": "Success",
  "data": {
    "msgid": "F15DE2491BEF86879BED"
  }
}
```

# Send Image
```
curl --request POST \
  --url http://server.ibnux.net:3000/api/v1/whatsapp/send/image \
  --header 'Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJkYXRhIjoiMDgxMjM0NTY3OCIsImV4cCI6MTYzMzUyNzk3NH0.4FLvNzZw3nwt93TxmnynQkVPYCaoRzUhdWD8ri6Fy74' \
  --header 'Content-Type: multipart/form-data' \
  --header 'content-type: multipart/form-data; boundary=---011000010111000001101001' \
  --form msisdn=628111204268@s.whatsapp.net \
  --form 'message=halo kang' \
  --form image=@/home/ibnuux/gambar.jpg
```

## RESULT
```
{
  "status": true,
  "code": 200,
  "message": "Success",
  "data": {
    "msgid": "F15DE2491BEF86879BED"
  }
}
```

# OTHER
Just see this code for other Endpoint
https://github.com/dimaskiddo/go-whatsapp-rest/blob/d6a00a048d98298c894d12e2853d9e1b0c9f71ab/internal/whatsapp/whatsapp.go#L60-L385

# Limitation
- You cannot use Whatsapp web, because this API will use it
- You need stand by phone for whatsapp, i use old android phone to make it always on
- sometimes it always logout, and need to scan QRcode, i make [script](https://gist.github.com/ibnux/e0f4a6411bcc5dfae1aef13ae9898b8c) to backup session in folder **/config/stores/**, if it empty, the script will copy it again, and its been 4 month never logout again when my phone disconnected from internet, because of indihome 😄

**Comment by dimaskiddo** [comment]
Thank you so much pak @ibnux, i hope this can help the others 😄  🍻

**Reply by ArivhaySoft** [reply]
Good sir @ibnux,

Thanks, **it just got easier**

don't forget to also thank you sir @dimaskiddo, the go-whatsapp-rest is really cool

**Comment by kuncolaxmono** [comment]
What's wrong about this? please help.. thanks

<img width="907" alt="Screen Shot 2021-07-28 at 13 06 24" src="https://user-images.githubusercontent.com/24507359/127272177-8b37a125-0790-44e5-b075-642a7c868a06.png">

**Reply by dimaskiddo** [reply]
@kuncolaxmono Please try to relogin first and the MSISDN not using @s.whatsapp.net just the phone number
